### ISSUE TITLE
=beginna =enda =fortnite

### ISSUE THREAD

**Issue by lizmat** [Raku/RakuDoc-GAMMA#33]
These are all pod / rakudoc markers that are **not** allowed in both the old implementation of pod6 as well as in rakudoc.

Why?  Because they share the root with `=begin`, `=end` and `=for`.

I was surprised to see this limitation in the legacy grammar as well.  And it looks like this limitation has been in there like forever.

So I guess the questions is: should this be fixed (I didn't see an immediate fix), or should this be documented?

**Comment by finanalyst** [comment]
@lizmat But you managed to fix `=formula`
Other than `=formula`, which is part of the new specification, all other forms should be caught because the spelling rules only allow lower case block names.

I have implemented a trap in the renderer for all lower case block names not specified (which would include the forms in the title), and create a RakuDoc warning, which is rendered at the end of the output, with some debug information.

The problem I found when `=formula` was failing is that the compiler does not provide much in the way of debug information, just a failure after the `=` of `=begin rakudoc`, which was confusing !!!!

**Comment by lizmat** [comment]
> But you managed to fix =formula

Yes, by adding a specific token for `=formula`.  And if we ever decide to allow `=beginna` or `'enda` as a specific directive, that can be fixed as well.  It's the general case of `=begin\w+`, `=end\w+` and `=for\w+` that are a problem.

> The problem I found when =formula was failing is that the compiler does not provide much in the way of debug information, just a failure after the = of =begin rakudoc, which was confusing !!!!

Agree.  The whole problem really is that from the compiler's point of view, *all* rakudoc is just whitespace.  This also appears to block localizationn, for that matter.

I guess my grammar foo is insufficient to fix this.  FWIW, I just tried an approach for `=begin\w+`, and that resulted in:
```
===SORRY!===
Invalid to edge 0 in NFA statelist
```
which is *very* low level and quite above my current paygrade.  So I abandoned this idea, at least for now.

**Comment by lizmat** [comment]
To illustrate it's all whitespace:
```raku
my

=foo bar

$a

=bar baz

=

=zippo

42

=donko

;

=fnoppie

say

=huh

$a
```
not only compiles, but also runs and says `42`

**Comment by finanalyst** [comment]
The current parser is doing enough  for a RakuDoc v2 renderer to work.
We already have Rakudo issues about better failure messages when RakuDoc does not compile, and being able to attach `# commentary` so that we ca get syntax highlighting via DEPARSE.

I think these are related to the above, so I suggest we close this issue here, or move to Rakudo.

**Comment by lizmat** [comment]
@finanalyst So you're saying the parser **should** be able to understand `=fortnite`.  And the fact that it doesn't, is a flaw in the implementation.

Since this bug has been in Rakudo like forever, I wonder if @thoughtstream would have any insights as to why this bug (still) exists.  If it is something that was found / discussed before, then I'd like to hear the outcome of that discussion!  :-)

**Comment by finanalyst** [comment]
@lizmat oh I misunderstood.
I thought that `=for\S+`, `=begin\S+`, and `=end\S+` caused compile time errors.
It does not matter for the Renderer I'm writing whether they are allowed or not. I have
```
given $ast.type ...
when # a list of named built-ins
when all($_.uniprops) ~~ / Lu / # to match semantic blocks
any($_.uniprops) ~~ / Lu / and any($_.uniprops) ~~ / Ll / # to match custom block
default { # generate an unknown built in block error }
```
If the compiler allows `=fornite` through, then it will be caught by the Renderer, and the document writer will get more debugging info than the compiler gives.

So, I suppose that perhaps the compiler should allow these forms through, and that it is for the renderer to deal with them.

Wouldnt that make the parser simpler to manage?

**Comment by thoughtstream** [comment]
Sorry, @lizmat, I have no insights into that long-standing issue with the Rakudo compiler’s handling of RakuDoc. I was not even aware of it.

**Comment by tbrowder** [comment]
My memory is not great, but I do remember contributing to some pod grammar toward NYI items, and one was a sweeping change to the giant parsing block where I moved all the pod-only parsing to the end to help work on pod.

Pod tests then were not very extensive, but did get added slowly. I could easily have made that =fortnite problem when going into a realm I didn't fully understood. One known failed attempt was the =defn construct when I couldn't grok how to handle beginning whitespace after the =defn.

**Comment by finanalyst** [comment]
@liz Just found another formula problem.
```
=begin rakudoc

=for formula :caption<Fabulous identity>
e^{i\pi}+1=0

=end rakudoc
```
generates
```
$ RAKUDO_RAKUAST=1 raku -c test-e.rakudoc 
===SORRY!=== Error while compiling /home/richard/development/RakuDoc-GAMMA/test-e.rakudoc
=begin may not be followed by directive 'for'
at /home/richard/development/RakuDoc-GAMMA/test-e.rakudoc:3
------> <BOL>⏏=begin formula :caption<Fabulous identit
```

**Comment by finanalyst** [comment]
@thoughtstream I am wondering whether we might compromise a bit.
`=formula` is a new addition to RakuDoc. `F<>` works without a problem.

The problem seems to be related to a deeper parsing problem. 

Perhaps we could substitute `=formula` for `=equation`?

Perhaps later, we could make `=formula` an alternative to `=equation`

**Comment by lizmat** [comment]
Fortunately, `=for formula` was easily fixed: https://github.com/rakudo/rakudo/commit/e2fa55b9b1

**Comment by thoughtstream** [comment]
Outstanding, @lizmat!

**Comment by jubilatious1** [comment]
> =begin\w+

Confused. Why are backslashes allowed in tags?

**Comment by finanalyst** [comment]
@jubilatious1 Is there any reason this has affected you? The `\W` is in the grammar. What am I not understanding?

**Comment by jubilatious1** [comment]
TL;DR
 _How does:_
 `=beginna`, `=enda`, or `=fortnite` __differ from__ `=for\S+`, `=begin\S+`, or `=end\S+`? 
 _And where/why would you use these?_
_Isn't it just simpler to eliminate?_

------------------

Hi @finanalyst , It has affected my by causing extreme confusion. 

My understanding is that RakuDoc was a re-imagination of old Perldocs that were thought to be in need of serious updating. So when I see syntax I don't understand I (occasionally) put up a note that says "I don't understand this".

If you (or @doomvox) could point me in the right direction--such as where or why you'd use `\W` in the grammar.

**Comment by doomvox** [comment]
@jubilatious1:  if you're going to be at the Raku Study Group meeting on Sunday (Nov 24th, 2024) we can talk about it then.  

The subject at hand is a limitation of the way raku docs are currently parsed, and Pod6 syntax includes things like:

    =for head1

Which is a tag followed by whitespace followed by a typename. One way of matching for that is to go for the word after some non-word stuff after the tag:  \W matches a non-word char, so \W+ is a way to suck up some whitespace.   So, the backslash isn't *in* the tag, it's in the grammar used to find the tag and the word after the tag.

**Comment by finanalyst** [comment]
@jubilatious1 @doomvox I'll try to join you on Sunday.
@doomvox has it  right. The problem that this issue was generated  by is a new block type for RakuDoc v2. 

We decided to include as a builtin the block `=formula` followed by a Latex definition of a formula. The problem was that the RakuAST parser was confusing `=formula` with `=for` (the normal Rakudo compiler does not recognise `=formula` at all). 

This problem has now been solved completely. This issue should be closed now. I only kept it open in case there was something we missed.

But if you want to create a new custom block,  it would need to be spelt with a mixture of upper and lower case letters, eg `=forThoseWhoGoToSleep`. This also would fall into the same category in that the beginning of the block name would be 'for'. It's not a good idea, but it is possible.

closing as resolved

**Comment by jubilatious1** [comment]
@doomvox 
@finanalyst 

> =for head1

Why is prefix `for` in RakuDoc, when Raku has postfix `for`?
```raku
=formula :caption<"Formula ++$"> for \@array-of-latex-formulas
```

**Comment by finanalyst** [comment]
@jubilatious1
1. Raku and RakuDoc are different languages. So `=for` is not `for`
2. `=for` serves is one of the three ways to mark the start of a block. We
have `=head`, `=for head` and `=begin head`. The latter two allow for
metadata such as `:!toc` which prevents the heading from being added to the
table of contents. The last allows for multiple paragraphs to be included
in the block.
3. Note: head is the same as head1 in RakuDoc V2, but is illegal in RakuDoc
v1. :toc has no meaning in RakuDoc v2, so is ignored.
4. The example you gave mixing `=for` and `for` is not legal.
5. Perhaps an easier way to think of the difference is that RakuDoc
elements are the same as white space or comments to the Raku code

**Comment by jubilatious1** [comment]
> 4. The example you gave mixing `=for` and `for` is not legal.

I don't recall mixing `=for` and `for`. 
I used `=formula` and `for`. 
This begs the question why the `=for ` block/tag does not require whitespace afterward. Or does it (now, V2)?

Thank you Richard.
